lib/Dialect/SV/Transforms/RTLLegalizeNames.cpp

```diff
@@ -0,0 +1,198 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transformation pass renames modules and variables to avoid conflicts
+// with keywords and other declarations.
+//
+//===----------------------------------------------------------------------===//
+
+#include "SVPassDetail.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "circt/Translation/ExportVerilog.h"
+
+using namespace circt;
+using namespace sv;
+using namespace rtl;
+
+//===----------------------------------------------------------------------===//
+// RTLLegalizeNamesPass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct RTLLegalizeNamesPass
+    : public sv::RTLLegalizeNamesBase<RTLLegalizeNamesPass> {
+  void runOnOperation() override;
+
+private:
+  bool anythingChanged;
+
+  /// Given a module name \p moduleName, return the updated name if it has been
+  /// previously renamed in the table moduleNameTable, else call
+  /// sanitizeName, to get a new name in case of conflict with
+  /// keywords.
+  StringRef getModuleName(StringAttr moduleName) {
+    auto entryIter = moduleNameTable.find(moduleName);
+    if (entryIter != moduleNameTable.end())
+      return entryIter->getSecond();
+    auto updatedName = sanitizeName(moduleName.getValue(), usedModuleNames,
+                                    nextGeneratedNameID);
+    moduleNameTable[moduleName] = updatedName;
+    return updatedName;
+  }
+
+  /// moduleNameTable keeps track of mappings from Module names to updated names
+  /// to resolve keyword conflicts.
+  llvm::DenseMap<Attribute, StringRef> moduleNameTable;
+
+  /// Set of used MLIR module names, to ensure unique names when renaming
+  /// keywords in module names.
+  llvm::StringSet<> usedModuleNames;
+
+  size_t nextGeneratedNameID = 0;
+};
+} // end anonymous namespace
+
+void RTLLegalizeNamesPass::runOnOperation() {
+  anythingChanged = false;
+  auto root = getOperation();
+
+  // We run in two phases. Phase one legalizes module and interface names in the
+  // module globally. Phase two then dives into these structures and performs
+  // local legalization on names that cannot leak out their enclosing module.
+
+  // Register the names of external modules which we cannot rename.
+  root.getBodyRegion().walk([&](Operation *op) {
+    if (auto extMod = dyn_cast<RTLModuleExternOp>(op)) {
+      auto verilogName = extMod.getVerilogModuleNameAttr();
+      getModuleName(verilogName);
+      // We expressely *do not* rename extern modules, where the user supplies
+      // the complementary definition from the outside.
+      // TODO: Maybe we should emit a diagnostic if this pass would rename an
```

Comment:
IMO, listing multiple extern modules with the same name should be illegal and result in an error. Listing the same extern module multiple times, however, should be allowed.

Reply:
Yeah that is a great point. This should go into verification somewhere though, right? Not necessarily into this pass?